File: sdk/typescript/src/hashing.ts
import { createHash } from "node:crypto";
import { base64UrlNoPad } from "./base64url.js";
import { canonicalizeToBytes } from "./jcs.js";

export function objectHash(objectType: string, objectValue: unknown): string {
  const canonical = canonicalizeToBytes(objectValue);
  const prefix = new TextEncoder().encode(`AICP1\0${objectType}\0`);
  const preimage = Buffer.concat([Buffer.from(prefix), Buffer.from(canonical)]);
  const digest = createHash("sha256").update(preimage).digest();
  return `sha256:${base64UrlNoPad(digest)}`;
}

export function messageHashFromBody(messageBody: Record<string, unknown>): string {
  return objectHash("message", messageBody);
}

File: interop/pairwise/release_artifacts/AICP-PAIRWISE-TCK-1.1.0/authority_root/reference/python/aicp_ref/hashing.py
from __future__ import annotations

import base64
import hashlib
from typing import Any

from .jcs import canonicalize_to_bytes


def b64url_no_pad(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def object_hash(object_type: str, obj: Any) -> str:
    canonical = canonicalize_to_bytes(obj)
    preimage = b"AICP1\0" + object_type.encode("utf-8") + b"\0" + canonical
    digest = hashlib.sha256(preimage).digest()
    return "sha256:" + b64url_no_pad(digest)


def message_hash_from_body(message_body_dict: dict[str, Any]) -> str:
    return object_hash("message", message_body_dict)
